py/mpstate.h

```diff
@@ -241,6 +241,16 @@ typedef struct _mp_state_vm_t {
 
     #if MICROPY_ENABLE_SCHEDULER
     volatile int16_t sched_state;
+
+    #if MICROPY_SCHEDULER_NODES
+    // These will usually point to statically allocated memory.  They are not
+    // traced by the GC.  They are assumed to be zero'd out before mp_init() is
+    // called (usually because this struct lives in the BSS).
```

Comment:
Where are these zeroed out on soft reset?

Reply:
They are zero'd an hard reset because they are in the BSS.

On soft reset they should not be zero'd because it's important that any outstanding scheduled callbacks are eventually called.  Eg if the BT driver has an event and schedules a callback, and it soft reset happens before the callback can be executed, then that callback will be executed once the runtime is restarted.

Reply:
Ah, makes sense now. This is just for scheduling "os"-type stuff, not for anything that uses the MicroPython runtime really.

Reply:
You could use the runtime -- BLE does -- but you need to be careful about it.